File: python/trebuchet.py
"""
https://adventofcode.com/2023/day/1
"""

def get_first_and_last_digit(sequence, alpha=False):
    """
    Returns the first and last digit of a given sequence.

    Args:
        sequence (str): The input sequence.
        alpha (bool, optional): Specifies whether to consider alphabetic characters as digits. 
                                Defaults to False.

    Returns:
        int: The first and last digit of the sequence.

    Example:
        >>> get_first_and_last_digit("abc123def")
        13
        >>> get_first_and_last_digit("two123def", alpha=True)
        23
    """
    sequence += " " * 4  # pad out

    digit_mapping = {
        "zero": "0",
        "one": "1",
        "two": "2",
        "three": "3",
        "four": "4",
        "five": "5",
        "six": "6",
        "seven": "7",
        "eight": "8",
        "nine": "9"
    }

    first_digit = None
    last_digit = None
    for i, char in enumerate(sequence):
        if char.isdigit():
            if first_digit is None:
                first_digit = char
            last_digit = char
        elif char.isalpha() and alpha:
            for key, value in digit_mapping.items():
                if char == key[0] and sequence[i:i + len(key)] == key:
                    if first_digit is None:
                        first_digit = value
                    last_digit = value

    if not first_digit:
        first_digit = ''
    if not last_digit:
        last_digit = ''

    return int(first_digit + last_digit)

if __name__ == "__main__":
    CALIB = 0
    ALPHA = 0

    with open("../input/trebuchet.txt", encoding='utf-8') as f:
        for line in f:
            CALIB += get_first_and_last_digit(line.strip(), False)
            ALPHA += get_first_and_last_digit(line.strip(), True)

    print(f'task 1: {CALIB}\ntask 2: {ALPHA}')

File: rust/src/trebuchet.rs
use std::fs::File;
use std::io::{self, BufRead};

fn get_first_and_last_digit(sequence: &str, alpha: bool) -> i32 {
    let digit_mapping = [
        ("zero", "0"),
        ("one", "1"),
        ("two", "2"),
        ("three", "3"),
        ("four", "4"),
        ("five", "5"),
        ("six", "6"),
        ("seven", "7"),
        ("eight", "8"),
        ("nine", "9"),
    ];

    let mut sequence = sequence.to_string();
    sequence.push_str("    "); // pad out

    let mut first_digit = None;
    let mut last_digit = None;

    for (i, char) in sequence.chars().enumerate() {
        if char.is_numeric() {
            if first_digit.is_none() {
                first_digit = Some(char);
            }
            last_digit = Some(char);
        } else if char.is_alphabetic() && alpha {
            for (key, value) in &digit_mapping {
                if char == key.chars().next().unwrap() && sequence[i..i+key.len()] == **key {
                    if first_digit.is_none() {
                        first_digit = Some(value.chars().next().unwrap());
                    }
                    last_digit = Some(value.chars().next().unwrap());
                    break;
                }
            }
        }
    }

    let result = format!("{}{}", first_digit.unwrap(), last_digit.unwrap());
    result.parse::<i32>().unwrap()
}

pub fn solve(input_directory: &str, file_name: &str, alpha: bool) -> io::Result<()> {

    let mut calibrations = 0;
    let file_path = format!("{}{}", input_directory, file_name);
    let file = File::open(file_path)?;

    for line in io::BufReader::new(file).lines() {
        let line = line?;
        calibrations += get_first_and_last_digit(line.trim(), alpha);
    }

    if alpha {
        println!("task 2: {}", calibrations);
    } else {
        println!("task 1: {}", calibrations);
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_first_and_last_digit_numeric() {
        let sequence = "12345";
        let alpha = false;
        let result = get_first_and_last_digit(sequence, alpha);
        assert_eq!(result, 15);
    }

    #[test]
    fn test_get_first_and_last_digit_alpha() {
        let sequence = "one two three four five";
        let alpha = true;
        let result = get_first_and_last_digit(sequence, alpha);
        assert_eq!(result, 15);
    }

    #[test]
    fn test_get_first_and_last_digit_mixed() {
        let sequence = "1 two 3 four five";
        let alpha = true;
        let result = get_first_and_last_digit(sequence, alpha);
        assert_eq!(result, 15);
    }
}